cmd/root.go

```diff
@@ -89,74 +90,97 @@ func rootCmd(o *options.Options) {
 
 	ctx := context.Background()
 	logger := sclog.NewLogger(sclog.ParseLevel(o.LogLevel))
-	repoURI, repoClient, ossFuzzRepoClient, ciiClient, vulnsClient, err := checker.GetClients(
-		ctx, o.Repo, o.Local, logger)
-	if err != nil {
-		log.Panic(err)
-	}
-	defer repoClient.Close()
-	if ossFuzzRepoClient != nil {
-		defer ossFuzzRepoClient.Close()
-	}
-
 	// Read docs.
 	checkDocs, err := docs.Read()
 	if err != nil {
 		log.Panicf("cannot read yaml file: %v", err)
 	}
 
-	var requiredRequestTypes []checker.RequestType
-	if o.Local != "" {
-		requiredRequestTypes = append(requiredRequestTypes, checker.FileBased)
-	}
-	if !strings.EqualFold(o.Commit, clients.HeadSHA) {
-		requiredRequestTypes = append(requiredRequestTypes, checker.CommitBased)
-	}
-	enabledChecks, err := policy.GetEnabled(pol, o.ChecksToRun, requiredRequestTypes)
-	if err != nil {
-		log.Panic(err)
-	}
+	switch o.Dependencydiff {
+	case "": /* Run the original scorecard checks on the repo. */
+		repoURI, repoClient, ossFuzzRepoClient, ciiClient, vulnsClient, err := checker.GetClients(
+			ctx, o.Repo, o.Local, logger)
+		if err != nil {
+			log.Panic(err)
+		}
+		defer repoClient.Close()
+		if ossFuzzRepoClient != nil {
+			defer ossFuzzRepoClient.Close()
+		}
 
-	if o.Format == options.FormatDefault {
-		for checkName := range enabledChecks {
-			fmt.Fprintf(os.Stderr, "Starting [%s]\n", checkName)
+		var requiredRequestTypes []checker.RequestType
+		if o.Local != "" {
+			requiredRequestTypes = append(requiredRequestTypes, checker.FileBased)
+		}
+		if !strings.EqualFold(o.Commit, clients.HeadSHA) {
+			requiredRequestTypes = append(requiredRequestTypes, checker.CommitBased)
+		}
+		enabledChecks, err := policy.GetEnabled(pol, o.ChecksToRun, requiredRequestTypes)
+		if err != nil {
+			log.Panic(err)
 		}
-	}
 
-	repoResult, err := pkg.RunScorecards(
```

Comment:
Would it be possible for the `GetDependencyDiffResult` API to have similar inputs as:
```
pkg.RunScorecards(
			ctx,
			repoURI,
			o.Commit,
			enabledChecks,
			repoClient,
			ossFuzzRepoClient,
			ciiClient,
			vulnsClient,
```
I mean all the client initialization part.

Looks like we're doing a lot of duplicated work / logic in the `GetDependencyDiffResult`, but it could be shared. Also, consumers like AllStar may want the same interface, since it seems to be working. I think it'd help to be consistent.

Wdut? Is there a blocker?

Reply:
IMO, the API `GetDependencyDiffResult` is designed for the caller to initialize a list of checks to run, and this should be a string array and is easy for the caller to specify. If we modify the API to have all those clients as input, it'll increase the complexity of using this API - users have to declare those corresponding clients to the checks to run, and we need to validate if the input clients - checks to run are matched.

I do agree that the current design of the client initialization part (see [here](https://github.com/ossf/scorecard/blob/main/dependencydiff/dependencydiff.go#L81-L83)) of `GetDependencyDiffResult` is a little duplicated. What I am doing in the func `initRepoAndClientByChecks` is, firstly I initialize all the clients using the original API `checker.GetClients` used in the scorecard check runner. Then, I return the required clients corresponding to the checks to run. 

Since the [current initialization](https://github.com/ossf/scorecard/blob/096cbd0622cc28ba321f920fd3ab833f5cdf81ab/cmd/root.go#L92-L135) for the scorecard check runner is to simply initialize and use all the clients if a valid `repoURI` is given, even if checks like `Fuzzing` or  `CII-Best-Practices` are not specified to run. Should I optimize the original API using mine?

Reply:
> IMO, the API `GetDependencyDiffResult` is designed for the caller to initialize a list of checks to run, this is a string array and should be easy for the caller to specify. If we modify the API to have all those clients as input, it'll increase the complexity of using this API - users have to declare those clients corresponding to the checks to run, and we need to validate if the input `clients` --- `checks to run` are matched (since they are from users).
> 

@jeffmendoza any insights on the API? I'm curious if we should follow the same pattern, since I suspect it must have been discussed before.

Reply:
Allstar needs to pass in a client/transport to handle authentication on its own. This is for when running scorecard checks as a library. I'm not sure we would plan to use this feature from Allstar directly.